src/allmydata/immutable/encode.py

```diff
@@ -694,3 +694,20 @@ def get_uri_extension_data(self):
         return self.uri_extension_data
     def get_uri_extension_hash(self):
         return self.uri_extension_hash
+
+    def get_uri_extension_size(self):
+        """
+        Calculate the size of the URI extension that gets written at the end of
+        immutables.
+
+        This may be done earlier than actual encoding, so e.g. we might not
+        know the crypttext hashes, but that's fine for our purposes since we
+        only care about the length.
+        """
+        params = self.uri_extension_data.copy()
+        assert params
+        params["crypttext_hash"] = b"\x00" * hashutil.CRYPTO_VAL_SIZE
+        params["crypttext_root_hash"] = b"\x00" * hashutil.CRYPTO_VAL_SIZE
+        params["share_root_hash"] = b"\x00" * hashutil.CRYPTO_VAL_SIZE
+        uri_extension = uri.pack_extension(params)
+        return len(uri_extension)
```

Comment:
So .. yea ... the downside to this implementation is that it's totally unclear what `params` contains (except it's not empty and it has the 3 values this code puts in it).

If it contains the wrong stuff then we might get the exact size wrong ... but at least we'll catch that with the change to `put_uri_extension` in the immutable layout code.  So it might be annoying to debug but it should still be safe with respect to writing shares.

It _ might_ make sense to strengthen the assert here - say, verify the keys of `params` are exactly the keys we know and expect in there.  However, that might be pointless polishing of some code that we pretty much already know we want to rewrite anyway.  So, skip it if you want.

Reply:
At minimum such an assertion would be handy for later refactoring, which would to do such validation, so I'll do that.